File: backend/src/main/java/com/smartcampus/booking/service/BookingService.java
package com.smartcampus.booking.service;

import com.smartcampus.auth.entity.Role;
import com.smartcampus.booking.dto.BookingRequest;
import com.smartcampus.booking.dto.BookingResponse;
import com.smartcampus.booking.entity.BookingStatus;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;

import java.time.LocalDateTime;
import java.util.List;
import java.util.Map;

public interface BookingService {

    BookingResponse create(BookingRequest request, Long userId);

    BookingResponse getById(Long id, Long viewerUserId, Role viewerRole);

    Page<BookingResponse> getAll(Pageable pageable);

    Page<BookingResponse> getByUser(Long userId, Pageable pageable);

    Page<BookingResponse> getFiltered(BookingStatus status, Long resourceId,
                                      LocalDateTime startDate, LocalDateTime endDate,
                                      Pageable pageable);

    BookingResponse approve(Long id, String adminComment);

    BookingResponse reject(Long id, String adminComment);

    BookingResponse cancel(Long id, Long userId, String reason);

    BookingResponse checkIn(Long id, Long userId);

    List<BookingResponse> getCalendar(Long resourceId, LocalDateTime start, LocalDateTime end);

    List<Map<String, Object>> getSuggestions(Long resourceId, java.time.LocalDate date, int durationMinutes);
}

File: frontend/src/services/bookingService.js
import api from './api';

export const bookingService = {
  getAll: (params) => api.get('/bookings', { params }),
  getMyBookings: (params) => api.get('/bookings/my', { params }),
  getById: (id) => api.get(`/bookings/${id}`),
  create: (data) => api.post('/bookings', data),
  update: (id, data) => api.put(`/bookings/${id}`, data),
  approve: (id, data) => api.patch(`/bookings/${id}/approve`, data),
  reject: (id, data) => api.patch(`/bookings/${id}/reject`, data),
  cancel: (id, reason) => api.patch(`/bookings/${id}/cancel`, reason ? { reason } : {}),
  checkIn: (id) => api.patch(`/bookings/${id}/check-in`),
  getCalendar: (params) => api.get('/bookings/calendar', { params }),
  getSuggestions: (params) => api.get('/bookings/suggestions', { params }),
  getMyWaitlistedBookings: (params) => api.get('/bookings/waitlist/my', { params }),
  leaveWaitlist: (id) => api.delete(`/bookings/${id}/waitlist`),
};
